bin/unit_test_script/run_unit_tests.py

```diff
@@ -0,0 +1,159 @@
+# running on bamboo ; trying to be as lights as posible
+#from shellwrap import color, unix
+import color # add color and verbose printing
+import unix # unix command calling
+
+# for a more descriptive run locally, remove these settings
+env = {"color": False, "verbose": color.VMode.ERROR}
+
+# There is nothing wrong with using a couple of globals vars to manage a bunch
+# of threads because this is way cleaner then doing it all in bash
+# pylint: disable=global-statement
+# pylint: disable=global-variable-not-assigned
+
+lock = Lock() # used to control when the globals are written to
+active_threads = 0 # pylint: disable=invalid-name
+total_time = 0 # pylint: disable=invalid-name
+total_jobs = 0 # pylint: disable=invalid-name
+work_list = [] # will hold a list of modules from lein to test
+base = os.getcwd()
+
+opt_out = ["cmr-dev-system",
+    "cmr-system-int-test",
+    "cmr-oracle-lib",
+    "cmr-mock-echo-app",]
+
+run_alone = ["cmr-common-app-lib", "cmr-indexer-app", "cmr-search-app"]
+
+# ##############################################################################
+
+def update_active_threads_locking(amount):
+    "Up the number of active threads, but make sure only one thread at a time can do this"
+    global active_threads
+    with lock:
+        active_threads += amount
+
+def update_total_time_locking(durration):
+    "Update the total time spent, but make sure only one thread at a time can do this"
+    global total_time
+    with lock:
+        total_time = total_time + durration
+
+def get_work_list():
+    "Get a dump of all the projects that lein manages"
+    raw_work_list = unix.pipe(["lein", "dump"])
+    list_of_projects = []
+    for item in raw_work_list.split("\n"):
+        item = item.strip(" ")
+        if item.startswith("cmr-") and (item not in opt_out) and (item not in run_alone):
+            list_of_projects.append(item[4:]) #remove cmr-
+        else:
+            print (f"skipping {item}")
+    return list_of_projects
+
+def worker(_context, id_number):
+    "Function for one thread, this is run many times."
+    global work_list, total_jobs, total_time, lock
+
+    color.cprint(color.tcode.green, f"Task {id_number} started.", environment=env)
+
+    update_active_threads_locking(1)
+
+    while 0 < len(work_list):
+        with lock:
+            total_jobs = total_jobs + 1
+            task = work_list.pop()
+
+        if task is None or len(task)<1:
+            color.cprint(color.tcode.red, f"restarting: {task}:{len(task)}",
+                verbose=color.VMode.ERROR, environment=env)
+            continue
+
+        color.cprint(color.tcode.white, f"+task {id_number} working on {task}.",
+            environment=env)
+
+        st = time.time()
+        # Run the external command in the task directory inside a try/except
+        # block, to ensure thread never dies
+        try:
+            os.chdir(base+"/"+task)
+            unix.pipe(["lein", "ci-utest"])
```

Comment:
Why is this a pipe?

Reply:
Since there is only one command being used, and the output of it isn't getting used, can we just use the subprocess.run directly?

Reply:
so this comes from a library function, I included the file so I would not need to do any pip actions, do you really want me to strip out all the functions because this tool does not use it?